platforms/core-configuration/groovy-support/src/main/java/org/gradle/api/internal/groovy/support/CompoundAssignmentTransformer.java

```diff
@@ -0,0 +1,192 @@
+            }
+            if (transformedExpr instanceof ClosureExpression) {
+                // Closure expression contains code, but ClosureExpression.transform doesn't descend into it.
+                ClosureExpression closureExpression = (ClosureExpression) transformedExpr;
+                closureExpression.visit(this);
+            }
+            return transformedExpr;
+        }
+
+        private Expression transformBinaryExpression(BinaryExpression expr) {
+            String operation = expr.getOperation().getText();
+            Integer compoundOpType = COMPOUND_TO_OPERATOR.get(operation);
+            if (compoundOpType != null) {
+                return transformCompoundAssignment(expr, compoundOpType);
+            }
+            return expr;
+        }
+
+        /**
+         * Rewrites {@code foo <OP>= bar} into {@code (foo = foo.forCompoundAssignment() <OP> (bar)).unwrap()}.
+         *
+         * @param original the original compound assignment expression
+         * @param compoundOperation the added operation of assignment ({@code OP})
+         * @return the transformed expression
+         */
+        private Expression transformCompoundAssignment(BinaryExpression original, int compoundOperation) {
+            Expression lhs = original.getLeftExpression();
+            Expression rhs = original.getRightExpression();
+
+            if (!isValidDestination(lhs)) {
+                // Skip array element assignments and the likes?
+                return original;
+            }
+
+            // Rewriting `foo <OP>= bar` into `foo = foo.forCompoundAssignment() <OP> (bar)`.
+            BinaryExpression assignment = withSourceLocationOf(original, new BinaryExpression(
+                ExpressionUtils.copyExpression(lhs), // Sharing the same node in different branches may cause other transforms to overwrite their metadata.
+                rewriteToken(original.getOperation(), Types.ASSIGN),
+                withSourceLocationOf(original, new BinaryExpression(
+                    applyForCompoundAssignment(lhs),
+                    rewriteToken(original.getOperation(), compoundOperation),
+                    rhs
+                ))
+            ));
+            // Final rewrite: unwrap the result
+            return withSourceLocationOf(original, applyToAssignmentResult(assignment));
+        }
+
+        /**
+         * Builds a new token from the original, keeping the source position.
+         *
+         * @param originalOperation the original operation token
+         * @param newType the new token type
+         * @return the new token
+         */
+        private Token rewriteToken(Token originalOperation, int newType) {
+            return Token.newSymbol(newType, originalOperation.getStartLine(), originalOperation.getStartColumn());
+        }
+
+        private Expression applyForCompoundAssignment(Expression expr) {
+            return callSupportMethodOn("forCompoundAssignment", expr);
```

Comment:
💭 Since it's just a string literal, I'd add a mention of `CompoundAssignmentExtensions` in a comment or javadoc